core/repo_status.py

```diff
@@ -0,0 +1,249 @@
+
+active_view = None  # type: Optional[sublime.View]
+current_token = {}  # type: Dict[RepoPath, str]
+
+
+class GsRegularStatusUpdaterEventListener(EventListener):
+    def on_new_async(self, view):
+        maybe_update_status_for_view(view)
+
+    # Sublime calls `on_post_save_async` events only on the primary view.
+    # We thus track the state of the `active_view` manually so that we can
+    # refresh the status bar of cloned views.
+
+    # Note: We listen for 'on_activated' bc we must draw on the main
+    # thread for the happy path to avoid visual yank on the status bar.
+    # We also only draw if the repo path is cached bc we must avoid
+    # doing expensive work on the main thread.
+    def on_activated_async(self, view):
+        global active_view
+        active_view = view
+
+        maybe_update_status_for_view(view)
+
+    def on_post_save_async(self, view):
+        global active_view
+        if active_view and active_view.buffer_id() == view.buffer_id():
+            maybe_update_status_for_view(active_view)
+        else:
+            maybe_update_status_for_view(view)
+
+
+class gs_update_status_bar(TextCommand):
+    def run(self, edit):
+        sublime.set_timeout_async(partial(maybe_update_status_for_view, self.view))
+
+
+def maybe_update_status_for_view(view):
+    """Record intent to update the status bar."""
+    if view_is_transient(view):
+        return
+
+    repo_path = repo_path_for_view(view)
+    if repo_path:
+        maybe_update_status_async(repo_path)
+
+
+def view_is_transient(view: sublime.View) -> bool:
+    """Return whether a view can be considered 'transient'.
+
+    For our purpose, transient views are 'detached' views or widgets
+    (aka quick or input panels).
+    """
+
+    # 'Detached' (already closed) views don't have a window.
+    window = view.window()
+    if not window:
+        return True
+
+    # Widgets are normal views but the typical getters don't list them.
+    group, index = window.get_view_index(view)
+    if group == -1:
+        return True
+
+    return False
+
+
+def maybe_update_status_async(repo_path):
+    # type: (RepoPath) -> None
+    with _lock:
+        current_token[repo_path] = token = uuid.uuid4().hex
+    sink = partial(update_status, repo_path)
+    pred = lambda: current_token.get(repo_path) == token
+    sublime.set_timeout_async(
+        partial(executor, pred, sink)
+    )
+
+
+def update_status(repo_path):
+    # type: (RepoPath) -> None
+    invalidate_token(repo_path)
+
+    git = make_git(repo_path)
+    status = fetch_status(git)
+    state.update_state(repo_path, status)
+
+
+def executor(pred, sink):
+    # type: (Thunk, Thunk) -> None
+    if pred():
+        sink()
+
+
+def invalidate_token(repo_path):
+    # type: (RepoPath) -> None
+    with _lock:
+        current_token.pop(repo_path, None)
+
+
+def make_git(repo_path):
+    # type: (RepoPath) -> GitCommand
+    git = GitCommand()
+    git.repo_path = repo_path
```

Comment:
We use the new setter `repo_path` here. It is not meant to be a general cache, bc we cache either on the `view.settings()` and of course in the new `repo_paths` global in `git_command`. What we need is a simple way to construct `GitCommand`s given a `repo_path` which is totally missing in the abstractions, super-classes. `make_git` itself should probably moved to `git_command`. Ideally, just `GitCommand(repo_path=...)` should be possible. 

Note that we currently store the `repo_path` in the `settings()` *only* for our fancy views because they otherwise have no `view.file_name()` anymore. After this PR, we don't use `settings` as a cache for *normal*, ordinary views the users uses to hack its code anymore. Normal views always get the value from the global cache because they have a `file_name()`. To the effect of that for the regular edge case of unsaved buffers we can still take the global cache.